src/contract.rs

```diff
@@ -27,14 +29,32 @@ pub fn init<S: Storage, A: Api, Q: Querier>(
     env: Env,
     msg: InitMsg,
 ) -> StdResult<InitResponse> {
+    // Check name, symbol, decimals
+    if !is_valid_name(&msg.name) {
+        return Err(StdError::generic_err(
+            "Name is not in the expected format (3-30 UTF-8 bytes)",
+        ));
+    }
+    if !is_valid_symbol(&msg.symbol) {
+        return Err(StdError::generic_err(
+            "Ticker symbol is not in expected format [A-Z]{3,6}",
+        ));
+    }
+    if msg.decimals > 18 {
+        return Err(StdError::generic_err("Decimals must not exceed 18"));
+    }
+
     let init_config = msg.config();
+    let admin = msg.admin.unwrap_or(env.message.sender);
+    let canon_admin = deps.api.canonical_address(&admin)?;
+
     let mut total_supply: u128 = 0;
     {
-        let mut balances = Balances::from_storage(&mut deps.storage);
```

Comment:
Not very, because `from_storage` itself doesn't perform storage access. Plus we need to do it like that because `store_mint` in the loop also takes `&mut deps.storage`, which would not compile if we had another mutable reference to storage at a larger scope (this is a good thing that maintains storage consistency)